File: src/cnft/indexer.ts
// Indexer orchestrator. Given an upstream + store, pulls signature
// history for a tree, walks each transaction, and applies every
// Bubblegum ix we understand to the store. Drives the step-2/3 pure
// machinery from real RPC data.
//
// Not a daemon — each call does one pass from the last-indexed
// signature forward and returns. Callers (handlers / CLI) decide when
// to re-run. That keeps this file free of timers, background tasks,
// and cancellation state — easier to reason about and to port.
//
// Failure strategy: one malformed tx or parse error never halts the
// scan. We log to stderr via `logUpstreamError`-style helpers and move
// on. Apply errors that represent genuinely-corrupt state (mint on
// unknown tree) are surfaced — they mean the user asked us to index a
// tree without first seeing its createTree, which we can't recover.

import type { Address } from "@solana/kit";
import type { UpstreamClient } from "../upstream.js";
import { applyEvent } from "./apply.js";
import { parseBubblegumInstruction } from "./parser.js";
import type { CnftStore } from "./store.js";
import { extractBubblegumIxs, type JsonRpcTransactionResponse } from "./tx-extract.js";

export interface IndexerDeps {
  upstream: UpstreamClient;
  store: CnftStore;
}

export interface IndexTreeOptions {
  /**
   * Cap on the number of signatures fetched in a single call. Safety
   * rail against accidentally backfilling hundreds of thousands of txs
   * on a wide production tree. Default 10_000. Set to null to uncap.
   */
  maxSignatures?: number | null;
  /**
   * Page size passed to getSignaturesForAddress. Solana's RPC caps this
   * at 1000. Default 1000.
   */
  pageSize?: number;
  /**
   * Side-channel for test instrumentation — called once per ix that
   * parsed successfully. Useful for assertions without exposing
   * internal counters on the return type.
   */
  onEventApplied?: (sig: string) => void;
}

export interface IndexTreeResult {
  /** Signatures we fetched and fully processed this call. */
  processed: number;
  /** Ixs parsed + applied across all processed txs. */
  applied: number;
  /** Ixs encountered but skipped (parse failures, unknown discriminators, etc.). */
  skipped: number;
  /** The newest signature we advanced the store cursor to, or null. */
  newestApplied: string | null;
}

interface SignatureEntry {
  signature: string;
  slot: number;
  err: unknown;
}

const DEFAULT_MAX_SIGNATURES = 10_000;
const DEFAULT_PAGE_SIZE = 1000;

/**
 * Do one incremental pass. Safe to call repeatedly — it picks up from
 * the store's `lastSignature` cursor each time, so callers can just
 * re-invoke to stay fresh.
 */
export async function indexTree(
  deps: IndexerDeps,
  tree: Address,
  options: IndexTreeOptions = {},
): Promise<IndexTreeResult> {
  const maxSigs = options.maxSignatures === null
    ? Infinity
    : (options.maxSignatures ?? DEFAULT_MAX_SIGNATURES);
  const pageSize = options.pageSize ?? DEFAULT_PAGE_SIZE;

  const cursor = await deps.store.getLastSignature(tree);
  const sigs = await fetchSignaturesUntil(deps.upstream, tree, cursor, pageSize, maxSigs);

  // fetchSignaturesUntil returns oldest-first — apply order matches
  // on-chain order, which is what state transitions need.
  const result: IndexTreeResult = {
    processed: 0,
    applied: 0,
    skipped: 0,
    newestApplied: null,
  };

  for (const sig of sigs) {
    if (sig.err !== null) {
      // On-chain failure — no state transitions to replay. Still
      // advance the cursor so the next pass skips this sig cheaply.
      await deps.store.setLastSignature(tree, sig.signature);
      result.newestApplied = sig.signature;
      result.processed++;
      continue;
    }

    const tx = await fetchTransaction(deps.upstream, sig.signature);
    if (!tx) {
      // Could not fetch the tx (pruned, transient error). Skip
      // without advancing cursor so a future pass can retry.
      result.skipped++;
      continue;
    }

    const ixs = extractBubblegumIxs(tx);
    for (const ix of ixs) {
      const parsed = parseBubblegumInstruction({
        data: ix.data,
        accounts: ix.accounts,
        noopEvent: ix.noopEvent,
      });
      if (!parsed.ok) {
        result.skipped++;
        continue;
      }
      if (parsed.value === null) {
        // Bubblegum ix we don't track yet (e.g. verifyCreator pre-step-4.5).
        result.skipped++;
        continue;
      }
      try {
        await applyEvent(deps.store, parsed.value);
        result.applied++;
        options.onEventApplied?.(sig.signature);
      } catch (err) {
        // Apply errors are rarely recoverable (mint on unknown tree,
        // etc). Log and continue rather than halt the whole scan.
        console.error(
          `[surfpool-helius cnft] applyEvent failed for ${sig.signature}:`,
          err instanceof Error ? err.message : err,
        );
        result.skipped++;
      }
    }

    await deps.store.setLastSignature(tree, sig.signature);
    result.newestApplied = sig.signature;
    result.processed++;
  }

  return result;
}

/**
 * Walk getSignaturesForAddress from newest toward `untilSig`, collect
 * everything (up to the cap), and return in chronological order.
 */
async function fetchSignaturesUntil(
  upstream: UpstreamClient,
  tree: Address,
  untilSig: string | null,
  pageSize: number,
  maxSigs: number,
): Promise<SignatureEntry[]> {
  const collected: SignatureEntry[] = [];
  let before: string | undefined;

  while (collected.length < maxSigs) {
    const remaining = maxSigs - collected.length;
    const limit = Math.min(pageSize, remaining);

    const params: Record<string, unknown> = { limit };
    if (before) params.before = before;
    if (untilSig) params.until = untilSig;

    const page = (await upstream.rpcCall("getSignaturesForAddress", [
      tree as unknown as string,
      params,
    ])) as SignatureEntry[] | null;

    if (!page || page.length === 0) break;
    collected.push(...page);
    if (page.length < limit) break;
    before = page[page.length - 1]!.signature;
  }

  // RPC returns newest-first. Reverse so we apply in chronological order.
  return collected.reverse();
}

async function fetchTransaction(
  upstream: UpstreamClient,
  signature: string,
): Promise<JsonRpcTransactionResponse | null> {
  try {
    const tx = (await upstream.rpcCall("getTransaction", [
      signature,
      { encoding: "json", maxSupportedTransactionVersion: 0, commitment: "confirmed" },
    ])) as JsonRpcTransactionResponse | null;
    return tx ?? null;
  } catch (err) {
    console.error(
      `[surfpool-helius cnft] getTransaction(${signature}) failed:`,
      err instanceof Error ? err.message : err,
    );
    return null;
  }
}

File: crates/service/src/cnft/indexer.rs
//! Indexer orchestrator. Given an upstream + store, pulls signature
//! history for a tree, walks each transaction, extracts Bubblegum ixs
//! (outer + inner, paired with their noop LeafSchemaEvents), parses
//! them into `CnftEvent`s, and applies each to the store.
//!
//! Not a daemon — each call does one incremental pass from the store's
//! last-indexed-signature cursor forward and returns. Callers choose
//! when to re-run. That keeps this file free of timers, cancellation
//! state, and background tasks — easier to reason about, easier to
//! port to other transport layers.
//!
//! Failure strategy: individual tx / ix failures never halt the scan.
//! We log via `tracing` and continue. Advance-and-skip is the policy
//! for known-failed txs (so a re-scan doesn't re-process them); skip-
//! without-advance is the policy for transient fetch errors (so they
//! get retried on the next pass).

use serde_json::json;
use thiserror::Error;
use tracing::warn;

use super::apply::{apply_event, ApplyError};
use super::parser::parse_bubblegum_instruction;
use super::store::{CnftStore, StoreError};
use super::tx_extract::{extract_bubblegum_ixs, RpcTransactionResponse};
use crate::upstream::{UpstreamClient, UpstreamError};

#[derive(Debug, Error)]
pub enum IndexError {
    #[error(transparent)]
    Store(#[from] StoreError),
    #[error(transparent)]
    Apply(#[from] ApplyError),
    #[error(transparent)]
    Upstream(#[from] UpstreamError),
    #[error("malformed signature batch from upstream: {0}")]
    MalformedSignatures(String),
}

pub type IndexResult<T> = Result<T, IndexError>;

#[derive(Debug, Clone)]
pub struct IndexTreeOptions {
    /// Cap on total signatures fetched in a single `index_tree` call.
    /// Safety rail against accidentally backfilling a 100k-tx tree in
    /// one shot. `None` = uncapped.
    pub max_signatures: Option<usize>,
    /// Page size for `getSignaturesForAddress`. Solana RPC caps this
    /// at 1000.
    pub page_size: usize,
}

impl Default for IndexTreeOptions {
    fn default() -> Self {
        Self {
            max_signatures: Some(10_000),
            page_size: 1000,
        }
    }
}

#[derive(Debug, Default, Clone, Copy, PartialEq, Eq)]
pub struct IndexTreeResult {
    /// Signatures we fetched and fully processed this call.
    pub processed: usize,
    /// Ixs parsed + applied across all processed txs.
    pub applied: usize,
    /// Ixs we encountered but skipped (parse errors, unknown
    /// discriminators, apply errors logged and moved past).
    pub skipped: usize,
}

/// One incremental pass. Safe to call repeatedly; picks up from the
/// store's `last_signature` cursor each time.
pub async fn index_tree<U, S>(
    upstream: &U,
    store: &S,
    tree: [u8; 32],
    options: &IndexTreeOptions,
) -> IndexResult<IndexTreeResult>
where
    U: UpstreamClient + ?Sized,
    S: CnftStore + ?Sized,
{
    let cursor = store.get_last_signature(&tree).await?;
    let sigs = fetch_signatures_until(upstream, &tree, cursor.as_deref(), options).await?;

    let mut result = IndexTreeResult::default();
    // RPC returns newest-first; we reverse to chronological so state
    // transitions replay in order.
    for entry in sigs.into_iter().rev() {
        if !entry.err_is_null() {
            // On-chain failure — no state to replay. Still advance the
            // cursor so a re-scan skips cheaply.
            store.set_last_signature(&tree, entry.signature.clone()).await?;
            result.processed += 1;
            continue;
        }

        let Some(tx) = fetch_transaction(upstream, &entry.signature).await else {
            // Skip without advancing — retry on next pass.
            result.skipped += 1;
            continue;
        };

        let ixs = extract_bubblegum_ixs(&tx);
        for ix in ixs {
            match parse_bubblegum_instruction(&ix.data, &ix.accounts, ix.noop_event.as_ref()) {
                Ok(Some(event)) => match apply_event(store, event).await {
                    Ok(()) => result.applied += 1,
                    Err(e) => {
                        warn!(sig = %entry.signature, err = %e, "apply_event failed; skipping");
                        result.skipped += 1;
                    }
                },
                Ok(None) => {
                    // Known Bubblegum ix we don't track (Redeem, V2 family, ...).
                    result.skipped += 1;
                }
                Err(e) => {
                    warn!(sig = %entry.signature, err = %e, "parse failed; skipping");
                    result.skipped += 1;
                }
            }
        }

        store.set_last_signature(&tree, entry.signature.clone()).await?;
        result.processed += 1;
    }

    Ok(result)
}

// ─── signature paging ──────────────────────────────────────────────

#[derive(Debug, Clone)]
struct SignatureEntry {
    signature: String,
    err: serde_json::Value,
}

impl SignatureEntry {
    fn err_is_null(&self) -> bool {
        self.err.is_null()
    }
}

async fn fetch_signatures_until<U: UpstreamClient + ?Sized>(
    upstream: &U,
    tree: &[u8; 32],
    until: Option<&str>,
    options: &IndexTreeOptions,
) -> IndexResult<Vec<SignatureEntry>> {
    let tree_b58 = bs58::encode(tree).into_string();
    let mut collected: Vec<SignatureEntry> = Vec::new();
    let mut before: Option<String> = None;
    let cap = options.max_signatures.unwrap_or(usize::MAX);

    while collected.len() < cap {
        let remaining = cap.saturating_sub(collected.len());
        let limit = options.page_size.min(remaining);
        let mut opts = serde_json::Map::new();
        opts.insert("limit".into(), json!(limit));
        if let Some(b) = &before {
            opts.insert("before".into(), json!(b));
        }
        if let Some(u) = until {
            opts.insert("until".into(), json!(u));
        }
        let params = json!([tree_b58, serde_json::Value::Object(opts)]);
        let raw = upstream.rpc_call("getSignaturesForAddress", params).await?;
        let page: serde_json::Value = serde_json::from_slice(&raw)
            .map_err(|e| IndexError::MalformedSignatures(e.to_string()))?;

        let entries = match page {
            serde_json::Value::Array(a) => a,
            serde_json::Value::Null => break,
            other => {
                return Err(IndexError::MalformedSignatures(format!(
                    "expected array, got {other:?}"
                )));
            }
        };
        if entries.is_empty() {
            break;
        }
        let batch_len = entries.len();
        for entry in entries {
            let signature = entry
                .get("signature")
                .and_then(|v| v.as_str())
                .ok_or_else(|| {
                    IndexError::MalformedSignatures("missing `signature` field".into())
                })?
                .to_string();
            let err = entry.get("err").cloned().unwrap_or(serde_json::Value::Null);
            collected.push(SignatureEntry { signature, err });
        }
        if batch_len < limit {
            break;
        }
        before = collected.last().map(|e| e.signature.clone());
    }

    Ok(collected)
}

async fn fetch_transaction<U: UpstreamClient + ?Sized>(
    upstream: &U,
    signature: &str,
) -> Option<RpcTransactionResponse> {
    let params = json!([
        signature,
        {
            "encoding": "json",
            "maxSupportedTransactionVersion": 0,
            "commitment": "confirmed"
        }
    ]);
    let raw = match upstream.rpc_call("getTransaction", params).await {
        Ok(r) => r,
        Err(e) => {
            warn!(sig = %signature, err = %e, "getTransaction failed; will retry next pass");
            return None;
        }
    };
    if raw.is_empty() || raw == b"null" {
        return None;
    }
    serde_json::from_slice::<RpcTransactionResponse>(&raw).ok()
}
